plugins/cue_points/base/batch/copyCuePoints/Engine/KMultiSourceCopyCuePointEngine.php

```diff
@@ -0,0 +1,54 @@
+<?php
+/**
+ * @package plugins.cuePoints
+ * @subpackage Scheduler
+ */
+class KMultiSourceCopyCuePointEngine extends KCopyCuePointEngine
+{
+    private $currentClip = null;
+
+    public function copyCuePoints()
+    {
+        $res = true;
+        foreach ($this->data->clipsDescriptionArray as $clipDescription) {
+            $this->currentClip = $clipDescription;
+            $res &= $this->copyCuePointsToEntry($clipDescription->sourceEntryId, $this->data->destinationEntryId);
+        }
+        return $res;
+    }
+
+    public function shouldCopyCuePoint($cuePoint)
+    {
+        $clipStartTime = $this->currentClip->startTime;
+        $clipEndTime = $clipStartTime + $this->currentClip->duration;
+        return is_null($cuePoint->endTime) || TimeOffsetUtils::onTimeRange($cuePoint->startTime,$cuePoint->endTime,$clipStartTime, $clipEndTime);
+    }
+
+    public function getCuePointFilter($entryId, $status = CuePointStatus::READY)
+    {
+        $filter = parent::getCuePointFilter($entryId, $status);
+        $filter->startTimeLessThanOrEqual = $this->currentClip->startTime + $this->currentClip->duration;
+        return $filter;
+    }
+
+    public function calculateCuePointTimes($cuePoint)
+    {
+        $clipStartTime = $this->currentClip->startTime;
+        $offsetInDestination = $this->currentClip->offsetInDestination;
+        $clipEndTime = $clipStartTime + $this->currentClip->duration;
+        $cuePointDestStartTime = TimeOffsetUtils::getAdjustedStartTime($cuePoint->startTime, $clipStartTime, $offsetInDestination);
+        $cuePointDestEndTime = null;
+        if ($cuePoint->endTime)
+            $cuePointDestEndTime = TimeOffsetUtils::getAdjustedEndTime($cuePoint->endTime, $clipStartTime ,$clipEndTime ,$offsetInDestination);

```

Comment:
Inside TimeOffsetUtils , if no cuepoint end time , use timeLineEndtime for calculations.